File: app/src/main/java/com/petproject/andy/bcstorage/data/BarcodeRepository.java
package com.petproject.andy.bcstorage.data;

import android.app.Application;
import android.arch.lifecycle.LiveData;
import android.os.AsyncTask;

import com.petproject.andy.bcstorage.data.database.Barcode;
import com.petproject.andy.bcstorage.data.database.BarcodeRoomDatabase;
import com.petproject.andy.bcstorage.data.database.DAO.BarcodeDAO;

import java.util.List;

public class BarcodeRepository
{

    private static final String INSERT_COMMAND = "insert";
    private static final String DELETE_COMMAND = "delete";
    private static final String DELETE_ALL_COMMAND = "delete-all";
    private static final String UPDATE_COMMAND = "update";
    private BarcodeDAO mBarcodeDao;
    private LiveData<List<Barcode>> mAllBarcodes;


    BarcodeRepository(Application application){
        BarcodeRoomDatabase db = BarcodeRoomDatabase.getDatabase(application);
        mBarcodeDao = db.barcodeDAO();
        mAllBarcodes = mBarcodeDao.getAllBarcodes();


    }
    //TODO: Make async
    public LiveData<List<Barcode>> getAllBarcodes(){
        return mAllBarcodes;
    }
    public LiveData<List<Barcode>> getSpecificBarcodes(String barcode){
        return mBarcodeDao.getSpecificBarcodes(barcode);
    }

    public void insert (Barcode barcode) {
        new DBAsyncTask(mBarcodeDao, INSERT_COMMAND).execute(barcode);
    }

    public void deleteAll(){
        new DBAsyncTask(mBarcodeDao, DELETE_ALL_COMMAND).execute();
    }

    public void delete (Barcode barcode) {
        new DBAsyncTask(mBarcodeDao, DELETE_COMMAND).execute(barcode);
    }

    public void update (Barcode barcode) {
        new DBAsyncTask(mBarcodeDao, UPDATE_COMMAND).execute(barcode);
    }

    private static class DBAsyncTask extends AsyncTask<Barcode, Void, Void> {

        private BarcodeDAO mAsyncTaskDao;
        private String command;

        DBAsyncTask(BarcodeDAO dao, String command) {
            mAsyncTaskDao = dao;
            this.command = command;
        }

        @Override
        protected Void doInBackground(final Barcode... params) {
            switch(command){
                case INSERT_COMMAND:
                    mAsyncTaskDao.insert(params[0]);
                    break;
                case DELETE_COMMAND:
                    mAsyncTaskDao.delete(params[0]);
                    break;
                case UPDATE_COMMAND:
                    mAsyncTaskDao.update(params[0]);
                    break;
                case DELETE_ALL_COMMAND:
                    mAsyncTaskDao.deleteAll();
            }

            return null;
        }
    }
}

File: app/src/main/java/com/petproject/andy/bcstorage/data/BarcodeRepository.kt
package com.petproject.andy.bcstorage.data

import android.app.Application
import android.arch.lifecycle.LiveData
import android.os.AsyncTask

import com.petproject.andy.bcstorage.data.database.Barcode
import com.petproject.andy.bcstorage.data.database.BarcodeRoomDatabase
import com.petproject.andy.bcstorage.data.database.dao.BarcodeDao

class BarcodeRepository internal constructor(application: Application) {
    private val mBarcodeDao: BarcodeDao
    //TODO: Make async
    val allBarcodes: LiveData<List<Barcode>>


    init {
        val db = BarcodeRoomDatabase.getDatabase(application)
        mBarcodeDao = db!!.barcodeDAO()
        allBarcodes = mBarcodeDao.allBarcodes


    }

    fun getSpecificBarcodes(barcode: String): LiveData<List<Barcode>> {
        return mBarcodeDao.getSpecificBarcodes(barcode)
    }

    fun insert(barcode: Barcode) {
        DBAsyncTask(mBarcodeDao, INSERT_COMMAND).execute(barcode)
    }

    fun deleteAll() {
        DBAsyncTask(mBarcodeDao, DELETE_ALL_COMMAND).execute()
    }

    fun delete(barcode: Barcode) {
        DBAsyncTask(mBarcodeDao, DELETE_COMMAND).execute(barcode)
    }

    fun update(barcode: Barcode) {
        DBAsyncTask(mBarcodeDao, UPDATE_COMMAND).execute(barcode)
    }

    private class DBAsyncTask internal constructor(private val mAsyncTaskDao: BarcodeDao, private val command: String) : AsyncTask<Barcode, Void, Void>() {

        override fun doInBackground(vararg params: Barcode): Void? {
            when (command) {
                INSERT_COMMAND -> mAsyncTaskDao.insert(params[0])
                DELETE_COMMAND -> mAsyncTaskDao.delete(params[0])
                UPDATE_COMMAND -> mAsyncTaskDao.update(params[0])
                DELETE_ALL_COMMAND -> mAsyncTaskDao.deleteAll()
            }
            return null
        }
    }

    companion object {

        private val INSERT_COMMAND = "insert"
        private val DELETE_COMMAND = "delete"
        private val DELETE_ALL_COMMAND = "delete-all"
        private val UPDATE_COMMAND = "update"
    }
}
